task/thermal/src/control.rs

```diff
@@ -259,51 +425,184 @@ impl<'a, B: BspT> ThermalControl<'a, B> {
                 self.post_failed_count = self.post_failed_count.wrapping_add(1);
             }
         }
-
-        // Prioritize returning errors, because they indicate that something is
-        // wrong with the sensors that are critical to the control loop.
-        last_err?;
-
-        Ok(worst_margin)
     }
 
     /// An extremely simple thermal control loop.
     ///
     /// Returns an error if the control loop failed to read critical sensors;
     /// the caller should set us to some kind of fail-safe mode if this
     /// occurs.
-    pub fn run_control(&mut self) -> Result<(), ThermalError> {
-        let mut r = self
-            .read_sensors()
-            .map_err(|_| ThermalError::DeviceError)?
-            .ok_or(ThermalError::NoReading)?;
-
-        r -= self.target_margin.0;
-        if r < 0.0f32 {
-            self.target_pwm = (self.target_pwm + 5).min(100);
-        } else if r < self.slow_band.0 {
-            self.target_pwm = (self.target_pwm + 1).min(100);
-        } else if r < self.dead_band.0 {
-            // No change
-        } else {
-            self.target_pwm = self.target_pwm.saturating_sub(1);
-        }
+    pub fn run_control(&mut self, now_ms: u64) -> Result<(), ThermalError> {
+        self.read_sensors(now_ms);
+
+        let control_result = match &mut self.state {
+            ThermalControlState::Boot { values } => {
+                let mut all_some = true;
+                let mut any_power_down = false;
+                let mut worst_margin = f32::MAX;
+                for (v, i) in values.iter().zip(self.bsp.inputs.iter()) {
+                    match v {
+                        Some(TemperatureReading::Valid { value, time_ms }) => {
+                            // Model the current temperature based on the last
+                            // reading and the worst-case slew rate.  This only
+                            // matters when samples are dropped; if we received
+                            // a reading on this control cycle, then time_ms ==
+                            // now_ms, so this becomes v.value (i.e. the most
+                            // recent reading).
+                            let temperature = value.0
+                                + (time_ms - now_ms) as f32
```

Comment:
nice.
Does this need to be capped to the sensor's maximum value?
We will want reports from thermal management on what outliers are causing fan pwm settings.

Reply:
I don't think we want to cap this to the sensor's maximum value: if the sensor's maximum value is 127°, we have a reading of 120° ten seconds ago, and the temperature could slew by 1°/sec, we want to know that the worst-case temperature is 130°, instead of clamping to 127°.

(Kinda a moot point, because we'll be in one of the fail-safe modes well before then!)